### ISSUE TITLE
feat: let `ctx.session` be undefined for missing sessions

### ISSUE THREAD

**Issue by KnorpelSenf** [grammyjs/grammY#713]
The fact that an error is thrown upon property access is really annoying. There is no good way of knowing whether session data exists, so we should changes this behaviour.

**Comment by KnorpelSenf** [comment]
We should probably still throw an error upon assignment to `ctx.session`

**Comment by roziscoding** [comment]
I get that it could be kind of annoying, but with filtering it'd be easy to get rid of updates missing the session object. We could even export a predicate ourselves, especially after we move sessions to a dedicated plugin

**Comment by KnorpelSenf** [comment]
What are you envisioning? `bot.filter(hasSession)`?

**Comment by roziscoding** [comment]
> What are you envisioning? `bot.filter(hasSession)`?

yep, something like that

**Comment by roziscoding** [comment]
> That feels a bit annoying, don't you think?

not really. the runtime check is already necessary for stuff that depend on the existence of a session key, well just reflect that in the types.

**Comment by KnorpelSenf** [comment]
I'd just like to find a way to make it unnecessary to check things at runtime. The current implementation doesn't need that and it would feel like a step back to require this from now on

**Comment by roziscoding** [comment]
> The current implementation doesn't need that
I might be missing something here, but I don't really get how this is true. isn't the whole purpose of this PQ to make the check easier? today, if you don't perform that runtime check, you risk getting an error

**Comment by KnorpelSenf** [comment]
I don't think you're missing something. It's just really annoying to be forced to check for `undefined`. It's also really annoying that there is not way to check for the presence of the object. In essence, in order to have an ergonomic API, we need `ctx.session` to be both present and optional at the same time. That's a contradiction. How do we resolve this?

- Compromise on the side of using `ctx.session` in the common case, and let `ctx.session` be optional
- Compromise on the side of checking the the presence of `ctx.session` in the rare case, and keep on throwing an error (=ignore this issue)
- Have `ctx.session` be present and still provide a way to check if it will throw, such as one of the following
  - `ctx.sessionKey === undefined`
  - `ctx.hasSession === true`
  - `hasSession(ctx) === true`
  - others
